File: deno/examples/network/network_manager.ts
// Example: Network Manager — agent identities, messaging, and peer discovery
// Demonstrates creating agent identities with capability cards, discovering
// peers via ManualDiscovery, constructing message envelopes (direct, broadcast,
// topic, reply), and using the PeerTable for routing decisions.
// Run: deno run deno/examples/agent-network/network_manager.ts

import {
  type AgentIdentity,
  broadcastEnvelope,
  BroadcastRouter,
  ContentRouter,
  createAgentIdentity,
  defaultAgentCard,
  directEnvelope,
  DirectRouter,
  displayTransportAddress,
  hasCapability,
  jsonPayload,
  ManualDiscovery,
  PeerTable,
  replyEnvelope,
  supportsProtocol,
  textPayload,
  topicEnvelope,
  type TransportAddress,
  withCorrelation,
  withTtl,
} from "@rullama/network";

async function main(): Promise<void> {
  console.log("=== Network Manager Example ===\n");

  // ---------------------------------------------------------------------------
  // 1. Create agent identities with capability cards
  // ---------------------------------------------------------------------------
  console.log("--- Agent Identities ---");

  const orchestrator: AgentIdentity = {
    ...createAgentIdentity("orchestrator"),
    agentCard: {
      ...defaultAgentCard(),
      capabilities: ["task-routing", "load-balancing"],
      supportedProtocols: ["mcp", "ipc"],
      endpoint: "unix:///tmp/orchestrator.sock",
      maxConcurrentTasks: 50,
      computeCapacity: 1.0,
    },
  };
  console.log(
    `  ${orchestrator.name} (id=${orchestrator.id})` +
      `\n    capabilities: ${
        JSON.stringify(orchestrator.agentCard.capabilities)
      }` +
      `\n    protocols:    ${
        JSON.stringify(orchestrator.agentCard.supportedProtocols)
      }` +
      `\n    endpoint:     ${orchestrator.agentCard.endpoint}`,
  );

  const workerA: AgentIdentity = {
    ...createAgentIdentity("worker-alpha"),
    agentCard: {
      ...defaultAgentCard(),
      capabilities: ["code-generation"],
      supportedProtocols: ["mcp", "ipc"],
      endpoint: "unix:///tmp/worker-a.sock",
      maxConcurrentTasks: 10,
      computeCapacity: 0.6,
    },
  };
  console.log(
    `  ${workerA.name} (id=${workerA.id})` +
      `\n    capabilities: ${JSON.stringify(workerA.agentCard.capabilities)}`,
  );

  const workerB: AgentIdentity = {
    ...createAgentIdentity("worker-beta"),
    agentCard: {
      ...defaultAgentCard(),
      capabilities: ["code-review", "testing"],
      supportedProtocols: ["mcp", "a2a"],
      endpoint: "tcp://127.0.0.1:9091",
      maxConcurrentTasks: 5,
      computeCapacity: 0.8,
    },
  };
  console.log(
    `  ${workerB.name} (id=${workerB.id})` +
      `\n    capabilities: ${JSON.stringify(workerB.agentCard.capabilities)}`,
  );
  console.log();

  // ---------------------------------------------------------------------------
  // 2. Check capabilities and protocol support
  // ---------------------------------------------------------------------------
  console.log("--- Capability & Protocol Checks ---");

  console.log(
    `  orchestrator has "task-routing": ${
      hasCapability(orchestrator.agentCard, "task-routing")
    }`,
  );
  console.log(
    `  workerA supports "mcp": ${supportsProtocol(workerA.agentCard, "mcp")}`,
  );
  console.log(
    `  workerB supports "ipc": ${supportsProtocol(workerB.agentCard, "ipc")}`,
  );
  console.log(
    `  workerB supports "a2a": ${supportsProtocol(workerB.agentCard, "a2a")}`,
  );
  console.log();

  // ---------------------------------------------------------------------------
  // 3. Peer discovery with ManualDiscovery
  // ---------------------------------------------------------------------------
  console.log("--- Peer Discovery ---");

  const discovery = ManualDiscovery.withPeers([workerA, workerB]);

  // Register orchestrator too
  await discovery.register(orchestrator);

  const peers = await discovery.discover();
  console.log(`  Discovered ${peers.length} peer(s):`);
  for (const peer of peers) {
    console.log(
      `    ${peer.name} — protocols: ${
        JSON.stringify(peer.agentCard.supportedProtocols)
      }, ` +
        `endpoint: ${peer.agentCard.endpoint ?? "none"}`,
    );
  }

  // Look up a specific peer
  const found = await discovery.lookup(workerA.id);
  console.log(`  Lookup workerA: ${found ? found.name : "not found"}`);
  console.log(`  Discovery protocol: ${JSON.stringify(discovery.protocol())}`);
  console.log();

  // ---------------------------------------------------------------------------
  // 4. Message envelope construction
  // ---------------------------------------------------------------------------
  console.log("--- Message Envelopes ---");

  // Direct message to worker-alpha
  const direct = directEnvelope(
    orchestrator.id,
    workerA.id,
    textPayload("Please generate a Rust HTTP handler"),
  );
  console.log(
    `  Direct:    sender=${direct.sender.slice(0, 8)}..., ` +
      `recipient=${JSON.stringify(direct.recipient)}, payload=Text("...")`,
  );

  // Broadcast to all peers
  const broadcast = broadcastEnvelope(
    orchestrator.id,
    textPayload("System: reloading config"),
  );
  console.log(
    `  Broadcast: sender=${broadcast.sender.slice(0, 8)}..., ` +
      `recipient=${JSON.stringify(broadcast.recipient)}`,
  );

  // Topic-addressed message
  const topic = topicEnvelope(
    orchestrator.id,
    "build-events",
    jsonPayload({ status: "success", duration_ms: 1234 }),
  );
  console.log(
    `  Topic:     sender=${topic.sender.slice(0, 8)}..., ` +
      `recipient=${JSON.stringify(topic.recipient)}`,
  );

  // Reply to the direct message
  const reply = replyEnvelope(
    direct,
    workerA.id,
    textPayload("Handler generated!"),
  );
  console.log(
    `  Reply:     sender=${reply.sender.slice(0, 8)}..., ` +
      `correlationId=${reply.correlationId?.slice(0, 8)}...`,
  );

  // TTL-limited message
  const ttlMsg = withTtl(
    broadcastEnvelope(orchestrator.id, textPayload("heartbeat")),
    3,
  );
  console.log(`  TTL msg:   ttl=${ttlMsg.ttl}`);

  // Correlation ID
  const correlated = withCorrelation(direct, "req-001");
  console.log(`  Correlated: correlationId=${correlated.correlationId}`);
  console.log();

  // ---------------------------------------------------------------------------
  // 5. PeerTable and routing
  // ---------------------------------------------------------------------------
  console.log("--- PeerTable & Routing ---");

  const peerTable = new PeerTable();

  const workerAAddr: TransportAddress = {
    type: "unix",
    path: "/tmp/worker-a.sock",
  };
  const workerBAddr: TransportAddress = {
    type: "tcp",
    address: "127.0.0.1:9091",
  };

  peerTable.upsert(workerA, [workerAAddr]);
  peerTable.upsert(workerB, [workerBAddr]);
  console.log(`  PeerTable: ${peerTable.length} peers`);

  for (const peer of peerTable.allPeers()) {
    const addrs = peerTable.getAddresses(peer.id) ?? [];
    const addrStrs = addrs.map(displayTransportAddress);
    console.log(`    ${peer.name} -> ${addrStrs.join(", ")}`);
  }

  // Topic subscriptions
  peerTable.subscribe(workerA.id, "build-events");
  peerTable.subscribe(workerB.id, "build-events");
  peerTable.subscribe(workerB.id, "review-events");
  console.log(
    `  Subscribers to "build-events": ${
      peerTable.subscribers("build-events").length
    }`,
  );
  console.log(
    `  Subscribers to "review-events": ${
      peerTable.subscribers("review-events").length
    }`,
  );
  console.log();

  // ---------------------------------------------------------------------------
  // 6. Route messages using different strategies
  // ---------------------------------------------------------------------------
  console.log("--- Routing Strategies ---");

  // Direct routing
  const directRouter = new DirectRouter();
  console.log(
    `  DirectRouter strategy: ${JSON.stringify(directRouter.strategy())}`,
  );
  const directAddrs = await directRouter.route(direct, peerTable);
  console.log(
    `    Route direct msg -> ${
      directAddrs.map(displayTransportAddress).join(", ")
    }`,
  );

  // Broadcast routing
  const broadcastRouter = new BroadcastRouter();
  console.log(
    `  BroadcastRouter strategy: ${JSON.stringify(broadcastRouter.strategy())}`,
  );
  const broadcastAddrs = await broadcastRouter.route(broadcast, peerTable);
  console.log(
    `    Route broadcast -> ${
      broadcastAddrs.map(displayTransportAddress).join(", ")
    }`,
  );

  // Content-based routing
  const contentRouter = new ContentRouter();
  console.log(
    `  ContentRouter strategy: ${JSON.stringify(contentRouter.strategy())}`,
  );
  const topicAddrs = await contentRouter.route(topic, peerTable);
  console.log(
    `    Route topic "build-events" -> ${
      topicAddrs.map(displayTransportAddress).join(", ")
    }`,
  );

  console.log("\nDone.");
}

await main();

File: crates/rullama-network/examples/network_manager.rs
//! Network Manager — agent registration, messaging, and event handling.
//!
//! Demonstrates:
//! - Building a `NetworkManager` with `NetworkManagerBuilder`
//! - Registering agents via `AgentIdentity` and `AgentCard`
//! - Peer discovery with `ManualDiscovery`
//! - Sending `MessageEnvelope` (direct, broadcast, topic)
//! - Subscribing to `NetworkEvent` and inspecting `ConnectionState`
//!
//! ```bash
//! cargo run -p rullama-network --example network_manager \
//!     --features "server,client,ipc-transport"
//! ```

use rullama_network::discovery::ManualDiscovery;
use rullama_network::{
    AgentCard, AgentIdentity, ConnectionState, MessageEnvelope, NetworkEvent,
    NetworkManagerBuilder, Payload, TransportType,
};

#[tokio::main]
async fn main() -> anyhow::Result<()> {
    println!("=== Network Manager Example ===\n");

    // -----------------------------------------------------------------------
    // 1. Create agent identities with capability cards
    // -----------------------------------------------------------------------
    println!("--- Agent Identities ---");

    let mut orchestrator = AgentIdentity::new("orchestrator");
    orchestrator.agent_card = AgentCard {
        capabilities: vec!["task-routing".into(), "load-balancing".into()],
        supported_protocols: vec!["mcp".into(), "ipc".into()],
        endpoint: Some("unix:///tmp/orchestrator.sock".into()),
        max_concurrent_tasks: Some(50),
        compute_capacity: Some(1.0),
        ..Default::default()
    };
    println!(
        "  {} (id={})\n    capabilities: {:?}\n    protocols:    {:?}\n    endpoint:     {:?}",
        orchestrator.name,
        orchestrator.id,
        orchestrator.agent_card.capabilities,
        orchestrator.agent_card.supported_protocols,
        orchestrator.agent_card.endpoint,
    );

    let mut worker_a = AgentIdentity::new("worker-alpha");
    worker_a.agent_card = AgentCard {
        capabilities: vec!["code-generation".into()],
        supported_protocols: vec!["mcp".into(), "ipc".into()],
        endpoint: Some("unix:///tmp/worker-a.sock".into()),
        max_concurrent_tasks: Some(10),
        compute_capacity: Some(0.6),
        ..Default::default()
    };
    println!(
        "  {} (id={})\n    capabilities: {:?}",
        worker_a.name, worker_a.id, worker_a.agent_card.capabilities,
    );

    let mut worker_b = AgentIdentity::new("worker-beta");
    worker_b.agent_card = AgentCard {
        capabilities: vec!["code-review".into(), "testing".into()],
        supported_protocols: vec!["mcp".into(), "a2a".into()],
        endpoint: Some("tcp://127.0.0.1:9091".into()),
        max_concurrent_tasks: Some(5),
        compute_capacity: Some(0.8),
        ..Default::default()
    };
    println!(
        "  {} (id={})\n    capabilities: {:?}",
        worker_b.name, worker_b.id, worker_b.agent_card.capabilities,
    );
    println!();

    // -----------------------------------------------------------------------
    // 2. Build the NetworkManager with manual discovery
    // -----------------------------------------------------------------------
    println!("--- Build NetworkManager ---");

    let discovery = ManualDiscovery::with_peers(vec![worker_a.clone(), worker_b.clone()]);

    let manager = NetworkManagerBuilder::new(orchestrator.clone())
        .add_discovery(Box::new(discovery))
        .event_buffer(128)
        .build();

    println!(
        "  Manager identity: {} (id={})",
        manager.identity().name,
        manager.identity().id
    );
    println!();

    // -----------------------------------------------------------------------
    // 3. Subscribe to network events
    // -----------------------------------------------------------------------
    println!("--- Subscribe to Events ---");
    let mut events = manager.subscribe();
    println!("  Event subscriber created");
    println!();

    // -----------------------------------------------------------------------
    // 4. Discover peers
    // -----------------------------------------------------------------------
    println!("--- Peer Discovery ---");

    let found = manager.discover_peers().await?;
    println!("  Discovered {} peer(s):", found.len());

    let peers = manager.peers().await;
    for peer in &peers {
        println!(
            "    {} — protocols: {:?}, endpoint: {:?}",
            peer.name, peer.agent_card.supported_protocols, peer.agent_card.endpoint
        );
    }
    println!();

    // -----------------------------------------------------------------------
    // 5. Drain PeerJoined events
    // -----------------------------------------------------------------------
    println!("--- Network Events ---");

    while let Ok(event) = events.try_recv() {
        match &event {
            NetworkEvent::PeerJoined(peer) => {
                println!("  Event: PeerJoined — {}", peer.name);
            }
            NetworkEvent::PeerLeft(id) => {
                println!("  Event: PeerLeft — {id}");
            }
            NetworkEvent::MessageReceived(env) => {
                println!("  Event: MessageReceived — from {}", env.sender);
            }
            NetworkEvent::ConnectionStateChanged { transport, state } => {
                println!("  Event: ConnectionStateChanged — {transport:?} -> {state:?}");
            }
            NetworkEvent::Error(e) => {
                println!("  Event: Error — {e:?}");
            }
        }
    }
    println!();

    // -----------------------------------------------------------------------
    // 6. Demonstrate message envelope construction
    // -----------------------------------------------------------------------
    println!("--- Message Envelopes ---");

    // Direct message to worker-alpha
    let direct = MessageEnvelope::direct(
        orchestrator.id,
        worker_a.id,
        Payload::Text("Please generate a Rust HTTP handler".into()),
    );
    println!(
        "  Direct:    sender={}, recipient={:?}, payload=Text(\"...\")",
        direct.sender, direct.recipient,
    );

    // Broadcast to all peers
    let broadcast = MessageEnvelope::broadcast(
        orchestrator.id,
        Payload::Text("System: reloading config".into()),
    );
    println!(
        "  Broadcast: sender={}, recipient={:?}",
        broadcast.sender, broadcast.recipient,
    );

    // Topic-addressed message
    let topic = MessageEnvelope::topic(
        orchestrator.id,
        "build-events",
        serde_json::json!({ "status": "success", "duration_ms": 1234 }),
    );
    println!(
        "  Topic:     sender={}, recipient={:?}",
        topic.sender, topic.recipient,
    );

    // Reply construction
    let reply = direct.reply(worker_a.id, Payload::Text("Handler generated!".into()));
    println!(
        "  Reply:     sender={}, correlation_id={:?}",
        reply.sender, reply.correlation_id,
    );

    // TTL-limited message
    let ttl_msg = MessageEnvelope::broadcast(orchestrator.id, "heartbeat").with_ttl(3);
    println!("  TTL msg:   ttl={:?}", ttl_msg.ttl);
    println!();

    // -----------------------------------------------------------------------
    // 7. Inspect connection state and transport type enums
    // -----------------------------------------------------------------------
    println!("--- Connection States & Transport Types ---");

    let states = [
        ConnectionState::Disconnected,
        ConnectionState::Connecting,
        ConnectionState::Connected,
        ConnectionState::Reconnecting,
    ];
    for state in &states {
        println!("  ConnectionState: {state:?}");
    }

    let transports = [
        TransportType::Ipc,
        TransportType::Remote,
        TransportType::Tcp,
        TransportType::A2a,
        TransportType::PubSub,
        TransportType::Custom("nats".into()),
    ];
    for t in &transports {
        println!("  TransportType:   {t:?}");
    }
    println!();

    // -----------------------------------------------------------------------
    // 8. Emit a synthetic event
    // -----------------------------------------------------------------------
    println!("--- Emit Synthetic Event ---");

    manager.emit(NetworkEvent::ConnectionStateChanged {
        transport: TransportType::Ipc,
        state: ConnectionState::Connected,
    });

    if let Ok(event) = events.try_recv() {
        println!("  Received: {event:?}");
    }

    println!("\nDone.");
    Ok(())
}
